### PR TITLE
Fix assertCount and the Count class

### LINKED ISSUE #3302: Expected size is wrong in assertCount error message
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 7.3.5
| PHP version         | 7.2.6
| Installation Method | Composer

I have this and the test passes:

```php
$this->assertCount( 6, $someGenerator );
```

I changed it from 6 to 5 to verify the test is being run

```php
$this->assertCount( 5, $someGenerator );
```

Got error

> Failed asserting that actual size 0 matches expected size 5.

While I expected to get

> Failed asserting that actual size 6 matches expected size 5.

### PR DESCRIPTION
**Pull request by JeroenDeDauw**
This is ready for review. Ping @sebastianbergmann 

Fixes #3302, builds upon and replaces https://github.com/sebastianbergmann/phpunit/pull/3304

Behavior changes with master:
* Error messages for non-rewindable Iterators are now shown correctly
* The `Count` class can now evaluate the same non-rewindable Iterator more than once
* Non-rewindable Iterators that are not Generators will no longer be rewound by `Count`, which for some Iterators causes an error

I took care to not change any other behavior. Though beware that the protected method `getCountOfGenerator` has been refactored away.

Tests added for the changed behavior and some clean up of existing tests, including adding assertions for existing behavior that was not tested yet.

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
Oh my. The CS checker insists on putting the methods in a suboptimal order :(

File: src/Framework/Constraint/Count.php [hunk 1]
```diff
@@ -56,64 +62,86 @@ protected function getCountOf($other): ?int
...
+    }
 
-            if (!$iterator instanceof Iterator) {
-                return \iterator_count($iterator);
-            }
+    private function getCountOfTraversable(Traversable $traversable): int
+    {
+        if ($this->iteratorCounts === null) {
+            $this->iteratorCounts = new \SplObjectStorage();
+        }
```
Comment (JeroenDeDauw) [inline src/Framework/Constraint/Count.php]: I do not understand how without this extra initialization the field can end up being null. Yet that is what apparently happens: https://travis-ci.org/sebastianbergmann/phpunit/jobs/434909673

How is this possible? It is set in the constructor. SameSize is calling `parent::__construct`. Nothing besides the constructor writes to the field and the field is private. Huh???

**Comment by JeroenDeDauw** [comment]
@sebastianbergmann @kubawerlos @guilliamxavier This is ready for review :)

**Review by guilliamxavier — CHANGES_REQUESTED** [review]
Thanks for the work! I have spotted a few issues (but I am not qualified to evaluate the whole picture)

File: src/Framework/Constraint/Count.php [hunk 2]
```diff
@@ -22,11 +21,17 @@ class Count extends Constraint
...
+     */
+    private $traversableCounts;
+
     public function __construct(int $expected)
     {
         parent::__construct();
 
-        $this->expectedCount = $expected;
+        $this->expectedCount     = $expected;
+        $this->traversableCounts = new \SplObjectStorage();
```
Comment (guilliamxavier) [inline src/Framework/Constraint/Count.php]: Question: since `$this->traversableCounts` is initialized as needed in `function getCountOfTraversable`, and is never used for the most common case of `array`|`Countable`, would it be preferable to _not_ initialize it here?

File: src/Framework/Constraint/Count.php [hunk 3]
```diff
@@ -116,4 +81,82 @@ protected function failureDescription($other): string
...
+            return $this->countIteratorWithoutChangingPositionIfPossible($traversable);
+        }
+
+        return $this->getCountOfTraversableThatIsNotIteratorOrIteratorAggregate($traversable);
+    }
+
+    private function countIteratorWithoutChangingPositionIfPossible(Iterator $iterator): int
+    {
+        $key   = $iterator->key();
+        $count = $this->getCountOfIterator($iterator);
```
Comment (guilliamxavier) [inline src/Framework/Constraint/Count.php]: Caution: original code uses custom iteration (without rewind) only for `Generator`, and `iterator_count` (which can rewind internally) for the rest, so I suspect this would change behavior (in edge cases). For example:

```php
$iterator = new \TestIterator([1, 2, 3]);
$iterator->next(); // this is the tricky bit

// which line below to uncomment, with original code? and with this PR?
//$n = 3;
//$n = 2;

Assert::assertCount($n, $iterator);
```

File: src/Framework/Constraint/Count.php [hunk 4]
```diff
@@ -116,4 +81,82 @@ protected function failureDescription($other): string
...
+
+        $this->traversableCounts->attach($iterator, $count);
+
+        if ($this->iteratorIsRewindable($iterator)) {
+            $this->traversableCounts->detach($iterator);
```
Comment (guilliamxavier) [inline src/Framework/Constraint/Count.php]: Maybe it is possible to `attach` only if needed, so that you don't need to `detach`

File: src/Framework/Constraint/Count.php [hunk 5]
```diff
@@ -116,4 +81,82 @@ protected function failureDescription($other): string
...
+            $this->rewindIterator($iterator, $key);
```
Comment (guilliamxavier) [inline src/Framework/Constraint/Count.php]: This originally was inside a condition `if ($key !== null) {`

File: src/Framework/Constraint/Count.php [hunk 6]
```diff
@@ -116,4 +81,82 @@ protected function failureDescription($other): string
...
+
+    private function iteratorIsRewindable(Iterator $iterator): bool
+    {
+        try {
+            $iterator->rewind();
+        } catch (\Exception $e) {
+            return false;
+        }
+
+        return !($iterator instanceof \NoRewindIterator);
```
Comment (guilliamxavier) [inline src/Framework/Constraint/Count.php]: Thoughts:
- Reorder `getCountOfIterator` before this function
- There is no need to try rewind if it is a `NoRewindIterator`
- If you cannot test whether an iterator is rewindable without actually trying to rewind it, maybe should you merge this function with `rewindIterator` below, so that the current position is unchanged (restored)

File: src/Framework/Constraint/Count.php [hunk 7]
```diff
@@ -116,4 +81,82 @@ protected function failureDescription($other): string
...
+    }
+
+    /**
+     * Returns the total number of iterations from a iterator.
+     * This will fully exhaust the iterator.
+     */
+    private function getCountOfIterator(Iterator $iterator): int
+    {
+        for ($countOfGenerator = 0; $iterator->valid(); $iterator->next()) {
```
Comment (guilliamxavier) [inline src/Framework/Constraint/Count.php]: `$countOfIterator`, or simpler (better?) just keep the original name `$count`

File: src/Framework/Constraint/Count.php [hunk 8]
```diff
@@ -116,4 +81,82 @@ protected function failureDescription($other): string
...
+        while ($iterator->valid() && $key !== $iterator->key()) {
+            $iterator->next();
+        }
+    }
+
+    private function getCountOfTraversableThatIsNotIteratorOrIteratorAggregate(Traversable $traversable): int
+    {
+        $count = \iterator_count($traversable);
+
+        $this->traversableCounts->attach($traversable, $count);
```
Comment (guilliamxavier) [inline src/Framework/Constraint/Count.php]: Maybe it is possible to factorize this `attach` with the other (i.e. directly in `function getCountOfTraversable`)

File: tests/_files/GeneratorTest.php [hunk 9]
```diff
@@ -0,0 +1,28 @@
+<?php
```
Comment (guilliamxavier) [inline tests/_files/GeneratorTest.php]: Is this file still needed? (if yes, then what about a more descriptive name?)

File: tests/unit/Framework/Constraint/CountTest.php [hunk 10]
```diff
@@ -142,4 +142,98 @@ public function testCountTraversable(): void
...
+
+            return;
+        }
+
+        $this->fail('An exception should have been thrown');
+    }
+
+    private function newPdoStatementWithThreeValues(): \Traversable
+    {
+        if (!\class_exists('\PDO')) {
```
Comment (guilliamxavier) [inline tests/unit/Framework/Constraint/CountTest.php]:
```php
        if (!\extension_loaded('pdo') || !\in_array('sqlite', \PDO::getAvailableDrivers())) {
```
according to https://github.com/sebastianbergmann/phpunit/blob/be6564ea49b96b9776c53f627a49683ec2687ce5/tests/end-to-end/regression/GitHub/1351.phpt

File: tests/unit/Framework/Constraint/CountTest.php [hunk 11]
```diff
@@ -142,4 +142,98 @@ public function testCountTraversable(): void
...
+            $this->markTestSkipped('derp');
```
Comment (guilliamxavier) [inline tests/unit/Framework/Constraint/CountTest.php]: What about `'PDO_SQLITE is required'` (still from https://github.com/sebastianbergmann/phpunit/blob/be6564ea49b96b9776c53f627a49683ec2687ce5/tests/end-to-end/regression/GitHub/1351.phpt)?

File: src/Framework/Constraint/Count.php (hunk 3 above)
Comment (JeroenDeDauw) [reply src/Framework/Constraint/Count.php]: This is tested by `testOnlyCountRemainingGeneratorElements`. I refactored that test to ascertain the original behavior and then made sure it stayed the same.

File: src/Framework/Constraint/Count.php (hunk 4 above)
Comment (JeroenDeDauw) [reply src/Framework/Constraint/Count.php]: As far as I can tell this is not possible, hence the approach taken.

File: src/Framework/Constraint/Count.php (hunk 5 above)
Comment (JeroenDeDauw) [reply src/Framework/Constraint/Count.php]: Many changes where made compared to the original. If you put in this condition now the tests will fail.

Comment (JeroenDeDauw) [reply src/Framework/Constraint/Count.php]: (If you want to see how it got removed, there are plenty of small commits to look at)

File: src/Framework/Constraint/Count.php (hunk 6 above)
Comment (JeroenDeDauw) [reply src/Framework/Constraint/Count.php]: > Reorder getCountOfIterator before this function

Done. That is some hard looking for things that can be improved there! In some way this is odd given this code is much higher quality than a lot of what I looked at while making these changes.

> If you cannot test whether an iterator is rewindable without actually trying to rewind it, maybe should you merge this function with rewindIterator below...

That was what the code was doing at some point. But then you have both the check, the rewind and a `traversableCounts->detach()` in there. Esp the detach is badly placed then.

File: src/Framework/Constraint/Count.php (hunk 7 above)
Comment (JeroenDeDauw) [reply src/Framework/Constraint/Count.php]: I did not change this code compared to the original

File: src/Framework/Constraint/Count.php (hunk 8 above)
Comment (JeroenDeDauw) [reply src/Framework/Constraint/Count.php]: I think that would make other things less nice and not be a good trade-off

File: tests/_files/GeneratorTest.php (hunk 9 above)
Comment (JeroenDeDauw) [reply tests/_files/GeneratorTest.php]: Probably good to have one integration test for the originally reported issue. Agree the name was not descriptive and changed it

File: tests/unit/Framework/Constraint/CountTest.php (hunk 11 above)
Comment (JeroenDeDauw) [reply tests/unit/Framework/Constraint/CountTest.php]: omg :D I forgot to put a real message there :D

File: src/Framework/Constraint/Count.php [hunk 12]
```diff
@@ -22,11 +21,16 @@ class Count extends Constraint
...
+     * @var null|\SplObjectStorage
+     */
+    private $traversableCounts;
+
     public function __construct(int $expected)
     {
         parent::__construct();
 
-        $this->expectedCount = $expected;
+        $this->expectedCount     = $expected;
```
Comment (guilliamxavier) [inline src/Framework/Constraint/Count.php]: leftover from intermediate commits, can be reverted back

File: tests/end-to-end/generator-failing-assert-count.phpt [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+F                                                                   1 / 1 (100%)
+
+Time: %s, Memory: %s
+
+There was 1 failure:
+
+1) Test\FailingGeneratorCountTest::testGenerator
+Failed asserting that actual size 3 matches expected size 4.
+
+%FailingGeneratorCountTest.php:%i
```
Comment (guilliamxavier) [inline tests/end-to-end/generator-failing-assert-count.phpt]:
```
%sFailingGeneratorCountTest.php:%i
```

(missing an `s` after the first `%`)
(that causes the Travis build to fail)

File: src/Framework/Constraint/Count.php (hunk 3 above)
Comment (guilliamxavier) [reply src/Framework/Constraint/Count.php]: I'm talking about *non*-`Generator` iterators:

```php
        $it = new \TestIterator([1, 2, 3]);
        $it->next();
        $this->assertTrue((new Count(3))->evaluate($it, '', true));
```

that was passing on master but is failing with this PR

File: src/Framework/Constraint/Count.php (hunk 6 above)
Comment (guilliamxavier) [reply src/Framework/Constraint/Count.php]: What about something like this?

```php
    private function getCountOfTraversable(Traversable $traversable): int
    {
        ...

        if ($this->traversableCounts->contains($traversable)) {
            return $this->traversableCounts[$traversable];
        }

        if ($traversable instanceof Iterator) {
            $count = $this->countIteratorWithoutChangingPositionIfPossible($traversable);
        } else {
            $count = \iterator_count($traversable);
        }

        $this->traversableCounts->attach($traversable, $count);

        return $count;
    }

    /**
     * Returns the total number of iterations from a iterator.
     * This will fully exhaust the iterator if it is not rewindable.
     */
    private function countIteratorWithoutChangingPositionIfPossible(Iterator $iterator): int
    {
        $key = $iterator->key();

        // Note: iterator_count calls rewind internally.
        try {
            $count = \iterator_count($iterator);
        } catch (\Exception $e) {
            // Fall back to manual iteration.
            for ($count = 0; $iterator->valid(); $iterator->next()) {
                ++$count;
            }

            return $count;
        }

        // Manually rewind $iterator to previous key, since iterator_count
        // moves pointer.
        if ($key !== null) {
            try {
                $iterator->rewind();
            } catch (\Exception $e) {
                return $count;
            }

            while ($iterator->valid() && $key !== $iterator->key()) {
                $iterator->next();
            }
        }

        return $count;
    }
```

It's simpler (I find) and keeps closer to the original code without needing to check for specific implementations like `NoRewindIterator` or even `Generator`. Maybe it can still be simplified or otherwise improved?

File: src/Framework/Constraint/Count.php (hunk 12 above)
Comment (JeroenDeDauw) [reply src/Framework/Constraint/Count.php]: Done

File: src/Framework/Constraint/Count.php (hunk 3 above)
Comment (JeroenDeDauw) [reply src/Framework/Constraint/Count.php]: Oh right. The Iterator case was also not tested >_>

Comment (JeroenDeDauw) [reply src/Framework/Constraint/Count.php]: Fixed now. Which allows getting rid of the ->detach()

**Review by guilliamxavier — APPROVED** [review]
This indeed seems an overall improvement (especially the tests) 👍

**Comment by JeroenDeDauw** [comment]
Thanks for the detailed review!

File: src/Framework/Constraint/Count.php [hunk 14]
```diff
@@ -56,64 +60,106 @@ protected function getCountOf($other): ?int
...
+            $this->moveIteratorToPosition($iterator, $key);
+        }
+
+        return $count;
     }
 
     /**
-     * Returns the total number of iterations from a generator.
-     * This will fully exhaust the generator.
+     * Returns the total number of iterations from a iterator.
```
Comment (guilliamxavier) [inline src/Framework/Constraint/Count.php]: Last comment (got lost when approving, sorry): this should now be "from a**n** iterator".
And maybe you can also improve "the total number of iterations" to the more accurate "the *remaining* number of iterations" or even "the *number of remaining iterations*"?

**Comment by epdenouden** [comment]
Will check this later today, too, because kittyyyyy

**Comment by sebastianbergmann** [comment]
This branch has conflicts that must be resolved. If the problem you wanted to address with this pull request still exists then please send a new pull request.

**Comment by JeroenDeDauw** [comment]
Yeah well not very inclined to put effort into that again after what happened here.
